injector/injector.go

```diff
@@ -0,0 +1,312 @@
+package injector
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/metrics"
+)
+
+const (
+	emptyAdmResponse = `<VAST version="3.0"><Ad><Wrapper><AdSystem>prebid.org wrapper</AdSystem><VASTAdTagURI><![CDATA[%s]]></VASTAdTagURI><Creatives></Creatives></Wrapper></Ad></VAST>`
+)
+
+type Injector interface {
+	InjectTracker(vastXML string, NURL string) string
+}
+
+type VASTEvents struct {
+	Errors                  []string
+	Impressions             []string
+	VideoClicks             []string
+	NonLinearClickTracking  []string
+	CompanionClickThrough   []string
+	LinearTrackingEvents    map[string][]string
+	NonLinearTrackingEvents map[string][]string
+	CompanionTrackingEvents map[string][]string
+}
+
+type TrackerInjector struct {
+	replacer macros.Replacer
+	events   VASTEvents
+	me       metrics.MetricsEngine
+	provider *macros.MacroProvider
+}
+
+var trimRunes = "\t\r\b\n "
+
+func NewTrackerInjector(replacer macros.Replacer, provider *macros.MacroProvider, events VASTEvents) *TrackerInjector {
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
```

Comment:
Does this TODO still need to be implemented.

Reply:
Will create a separate pr for prometheus changes.